### ISSUE TITLE
Doesn't work with npm@7 and above [work almost finished, see @next tag on npm]

### ISSUE THREAD

**Issue by mgdodge** [naugtur/npm-audit-resolver#34]
Using nvm to run node 14, but update npm to latest version (v 7.5.3). Project already has audit-resolve.json file. Running `check-audit` results in the following:

```
TypeError: Cannot read property 'length' of undefined
    at /home/********/node_modules/npm-audit-resolver/check.js:22:45
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
npm ERR! code 2
npm ERR! path /home/********
npm ERR! command failed
npm ERR! command sh -c check-audit

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/********/.npm/_logs/2021-02-09T00_48_37_076Z-debug.log
```

That log file contains the following:

```0 verbose cli [
0 verbose cli   '/home/********/.nvm/versions/node/v14.15.3/bin/node',
0 verbose cli   '/home/********/.nvm/versions/node/v14.15.3/bin/npm',
0 verbose cli   'run',
0 verbose cli   'check-audit'
0 verbose cli ]
1 info using npm@7.5.3
2 info using node@v14.15.3
3 timing config:load:defaults Completed in 1ms
4 timing config:load:file:/home/********/.nvm/versions/node/v14.15.3/lib/node_modules/npm/npmrc Completed in 1ms
5 timing config:load:builtin Completed in 1ms
6 timing config:load:cli Completed in 2ms
7 timing config:load:env Completed in 0ms
8 timing config:load:file:/home/********/.npmrc Completed in 0ms
9 timing config:load:project Completed in 1ms
10 timing config:load:file:/home/********/.npmrc Completed in 1ms
11 timing config:load:user Completed in 1ms
12 timing config:load:file:/home/********/.nvm/versions/node/v14.15.3/etc/npmrc Completed in 0ms
13 timing config:load:global Completed in 0ms
14 timing config:load:cafile Completed in 1ms
15 timing config:load:validate Completed in 0ms
16 timing config:load:setUserAgent Completed in 0ms
17 timing config:load:setEnvs Completed in 1ms
18 timing config:load Completed in 8ms
19 verbose npm-session 298dde723420661e
20 timing npm:load Completed in 17ms
21 timing command:run-script Completed in 2324ms
22 verbose stack Error: command failed
22 verbose stack     at ChildProcess.<anonymous> (/home/********/.nvm/versions/node/v14.15.3/lib/node_modules/npm/node_modules/@npmcli/promise-spawn/index.js:64:27)
22 verbose stack     at ChildProcess.emit (events.js:315:20)
22 verbose stack     at maybeClose (internal/child_process.js:1048:16)
22 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)
23 verbose pkgid ********@0.1.0
24 verbose cwd /home/********
25 verbose Linux 4.19.104-microsoft-standard
26 verbose argv "/home/********/.nvm/versions/node/v14.15.3/bin/node" "/home/********/.nvm/versions/node/v14.15.3/bin/npm" "run" "check-audit"
27 verbose node v14.15.3
28 verbose npm  v7.5.3
29 error code 2
30 error path /home/********
31 error command failed
32 error command sh -c check-audit
33 verbose exit 2
```

Have a variety of projects, some using `check-audit` in commit hooks like this, others using it in CI build hooks - really causing a lot of issues!

**Comment by naugtur** [comment]
Great to hear you're using npm-audit-resolver.
I've built it in my hobby time and am working to get it built into nom and other package managers. It's a slow moving initiative tho.

About npm7 - looks like the format of audit json changed.

**Comment by mgdodge** [comment]
Found this enlightening - https://uko.codes/dealing-with-npm-v7-audit-changes

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by hanvyj** [comment]
I get a similar issue with npm 6.13.4:

```
TypeError: Cannot read property 'length' of undefined
    at pkgFacade.getAudit.then.input (...\node_modules\npm-audit-resolver\check.js:23:45)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

**Comment by naugtur** [comment]
Thanks for the link.

The output from npm7 makes some core audit-resolver features impossible. 

Resolver is taking care of the dependency tree paths in which things get installed so that it reports the vulnerability as not-ignored if it finds it in a different dependency path (a new package starts using the ignored thing so maybe you no longer want to ignore it).
npm7 won't let me do that.
Gonna have to talk to npm folks...

**Comment by mgdodge** [comment]
Frustrating. Guess I'll be sticking to npm 6 for now. This utility has been invaluable to a number of projects I have been working on, so I'm not migrating them to npm 7 until I can ensure a similar audit experience.

**Comment by mgdodge** [comment]
Is it worth doing a patch-level release to update the README and tell people it won't work properly with npm 7?

**Comment by naugtur** [comment]
yup, I'll do that.

I'm also considering getting a limited version out without the path protection, but wanted to talk to npm folks first.

**Comment by naugtur** [comment]
I might have a way. 
@mgdodge or anynone else - would you be willing to share a result of `npm audit --json` on your repositories for npm6 and npm7 for me to explore?

**Comment by mgdodge** [comment]
I'll see if I can find a project that can easily be converted to a minimal example, then create a sample repo.

**Comment by naugtur** [comment]
@mgdodge All I need is the npm audit --json output from npm6 and npm7 on your project, so if you feel comfortable sharing with me (doesn't have to be public) the list of your vulns at a point in time, It'd be enough.

**Comment by mgdodge** [comment]
Here is a gist with two example `npm audit --json` outputs:

https://gist.github.com/mgdodge/9d0601e76198af918f2542a5503c9f9e

The repo this is based off of is actually for a vue bug I reported a few months ago - https://github.com/mgdodge/rollup-plugin-vue-treeshake-bug-vue3. If you want/need a full checkout of the project to try running `npm ci`, `npm update` etc. against, these are from the vue2 branch of that repo.

**Comment by naugtur** [comment]
Thanks. 
MY issue with output from npm7 is evident in your example. Look at y18n - in npm6 output you can find paths where the package is at. If you decide none of these affect you, you can ignore it and audit-resolver will NOT ignore it if it shows up in a new location in dependency tree. 
Output from npm7 doesn't have that information, so ignoring it once will ignore it even if the vulnerable dependency gets pulled by another package that might actually call it from your production code.

Protecting against this was the main reason why audit-resolver was created. Ignoring was possible in much simpler ways.

**Comment by jakawell** [comment]
Could that information be derived from the `package-lock.json` file? I know that would a lot more complex...

**Comment by naugtur** [comment]
@jakawell yes, that's the only option left, but correlating it with the right info from audit and the resolve.json file is going to be a lot of work.

**Comment by mgdodge** [comment]
Not to rain on the parade, but the `package-lock.json` file also underwent changes with npm7. I don't know the details, but [this blog post](https://nitayneeman.com/posts/catching-up-with-package-lockfile-changes-in-npm-v7/) as well as [this one](https://blog.npmjs.org/post/621733939456933888/npm-v7-series-why-keep-package-lockjson.html) might provide some insight. So, not only will there be a need to support npm6 and npm7 audit json output, but possibly package-lock.json output as well. 🤷‍♂️

**Comment by joebowbeer** [comment]
@mgdodge to be optimistic, I hope the new v7 format and hidden file make it easier for the audit resolver to work.

@naugtur what are your expectations regarding integration of the audit resolver with npm audit? Is there a statement of intent from npm or anything on the roadmap?

**Comment by adamzerella** [comment]
Bumping for an update on this issue, I'm also experiencing it with `npm 7`

**Comment by naugtur** [comment]
Update - I think I now know how to recover the missing information by re-indexing the current format. I'll give it a try. 
Anyone willing to also share their npm7 audit --json results?

**Comment by naugtur** [comment]
@mgdodge well, the only limitation is what you feel comfortable sharing. It's a list of vulnerabilities in an app afterall. Send whatever - 3 examples I didn't artificially create is better than 1 :)

**Comment by bytestream** [comment]
@naugtur would adding `npm@6` to this repository's `package.json` temporarily resolve this (until npm 7 is supported)?

I've ran into this after updating `@semantic-release/npm` which now requires `npm@7`. It breaks `check-audit` even though I have npm `6.14.9` installed

**Comment by naugtur** [comment]
Yes, installing npm 6 helps. 
I first saw this issue because of semantic release. I ended up putting semanticrelease in dev dependencies and running audit from global install or npx while installing with --production

**Comment by naugtur** [comment]
Good news @here 

I've published an early version with npm7 support. Individual package fixing is not implemented, but otherwise it should work. 

Check the v3 prerelease by installing `npm i npm-audit-resolver@next` or `npm i npm-audit-resolver@3.0.0-0`

**Comment by naugtur** [comment]
Yes, npm decided to stop scrambling private packages. But thanks for cross-checking. I don't have capacity to do solid testing.

Let me know whatever you notice!

**Comment by jamesots** [comment]
Because it's using a git url for audit-resolve-core, I'm unable to get the prerelease to work in Azure pipelines, as I get this error when it runs `npm ci`:

```
npm ERR! Error while executing:
npm ERR! /usr/bin/git ls-remote -h -t ssh://git@github.com/naugtur/audit-resolve-core.git
npm ERR! 
npm ERR! Warning: Permanently added the RSA host key for IP address '140.82.121.4' to the list of known hosts.
npm ERR! git@github.com: Permission denied (publickey).
npm ERR! fatal: Could not read from remote repository.
npm ERR! 
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.
npm ERR! 
npm ERR! exited with error code: 128
```

**Comment by joebowbeer** [comment]
@jamesots a lot of people are watching this issue: https://github.com/npm/cli/issues/2610

Related?

**Comment by naugtur** [comment]
I can switch to a proper package rc tonight to unblock if you promise to report back with feedback.
Pinkie promise? 😜

**Comment by naugtur** [comment]
3.0.0-1 has a regular dependency instead of a github url.

**Comment by jamesots** [comment]
It now installs and runs in Azure pipelines. But I now have a different error: 

```
>>>> npm audit --json
>>>> npm ls --depth=0 --json
RangeError: Maximum call stack size exceeded
    at Array.forEach (<anonymous>)
    at indexPaths (C:\path\node_modules\npm-audit-resolver\src\pkgmanagers\npm.js:44:18)
    at C:\path\node_modules\npm-audit-resolver\src\pkgmanagers\npm.js:46:17
    at Array.forEach (<anonymous>)
    at indexPaths (C:\path\node_modules\npm-audit-resolver\src\pkgmanagers\npm.js:44:18)
    at C:\path\node_modules\npm-audit-resolver\src\pkgmanagers\npm.js:46:17
    at Array.forEach (<anonymous>)
    at indexPaths (C:\path\node_modules\npm-audit-resolver\src\pkgmanagers\npm.js:44:18)
    at C:\path\node_modules\npm-audit-resolver\src\pkgmanagers\npm.js:46:17
    at Array.forEach (<anonymous>)
```

However, it's also happening with 3.0.0-0, so I guess there's a different audit response coming back from npm since I last tried.

I put some debug logging in, and there appears to be a circular dependency between gatsby and gatsby-plugin-utils which is putting npm-audit-resolver into a spin:

```
    "gatsby": {
      "name": "gatsby",
      "severity": "high",
      "via": [
        "autoprefixer",
        "gatsby-cli",
        "gatsby-plugin-utils",
        "optimize-css-assets-webpack-plugin",
        "postcss-flexbugs-fixes",
        "postcss-loader",
        "webpack-dev-server"
      ],
      "effects": [
        "gatsby-plugin-manifest",
        "gatsby-plugin-offline",
        "gatsby-plugin-utils",
        "gatsby-source-filesystem",
        "gatsby-transformer-sharp"
      ],
      "range": ">=2.0.118",
      "nodes": [
        "node_modules/gatsby"
      ],
      "fixAvailable": {
        "name": "gatsby",
        "version": "3.6.2",
        "isSemVerMajor": true
      }
    },
    "gatsby-plugin-utils": {
      "name": "gatsby-plugin-utils",
      "severity": "moderate",
      "via": [
        "gatsby"
      ],
      "effects": [
        "gatsby",
        "gatsby-plugin-manifest"
      ],
      "range": "0.2.27-qod.19 - 0.2.27-qod.21 || 0.2.31-query-modules-debug-missing-sq.23 || 0.2.35 - 0.10.0-next.0",
      "nodes": [
        "node_modules/gatsby-plugin-utils"
      ],
      "fixAvailable": {
        "name": "gatsby",
        "version": "3.6.2",
        "isSemVerMajor": true
      }
    },
```

**Comment by naugtur** [comment]
Love it!
Could you please send me your package.json and package-lock so I can create a full repro case?
Doesn't have to be in public if you don't want. You can email it to me too. Same login but at gmail

**Comment by jamesots** [comment]
I've created a very small test which is showing the error. I've put it in a gist: https://gist.github.com/jamesots/ac65692f601c7afb2510241da02a45d9

I included the output from npm audit too.

**Comment by naugtur** [comment]
@jamesots try v3.0.0-2 
I thought it'll be more fun to fix. turns out already had everything in place to track down the cycles. 
I used your audit output in a test. 

Works for me. Let me know what else you find! :)

**Comment by DavidJFowler** [comment]
I can't get v3.0.0-2 or v3.0.0-1 to work. v3.0.0-0 is OK.

I get this error on running check-audit:

```
internal/modules/cjs/loader.js:888
  throw err;
  ^

Error: Cannot find module 'concat-stream'
Require stack:
- D:\npm\node_modules\npm-audit-resolver\src\promiseCommand.js
- D:\npm\node_modules\npm-audit-resolver\src\pkgFacade\index.js
- D:\npm\node_modules\npm-audit-resolver\check.js
[90m    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:885:15)[39m
[90m    at Function.Module._load (internal/modules/cjs/loader.js:730:27)[39m
[90m    at Module.require (internal/modules/cjs/loader.js:957:19)[39m
[90m    at require (internal/modules/cjs/helpers.js:88:18)[39m
    at Object.<anonymous> (D:\npm\node_modules\[4mnpm-audit-resolver[24m\src\promiseCommand.js:2:16)
[90m    at Module._compile (internal/modules/cjs/loader.js:1068:30)[39m
[90m    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1097:10)[39m
[90m    at Module.load (internal/modules/cjs/loader.js:933:32)[39m
[90m    at Function.Module._load (internal/modules/cjs/loader.js:774:14)[39m
[90m    at Module.require (internal/modules/cjs/loader.js:957:19)[39m {
  code: [32m'MODULE_NOT_FOUND'[39m,
  requireStack: [
    [32m'D:\\npm\\node_modules\\npm-audit-resolver\\src\\promiseCommand.js'[39m,
    [32m'D:\\npm\\node_modules\\npm-audit-resolver\\src\\pkgFacade\\index.js'[39m,
    [32m'D:\\npm\\node_modules\\npm-audit-resolver\\check.js'[39m
  ]
}
```
Windows 10 1909 Build 18363.1556
Node 14.17.0
Npm 7.15.1

**Comment by naugtur** [comment]
This happened to me once before but I thought it was because if the mess I made in local node_modules
Wonder why this doesn't happen to everyone...

I'll fix it in the evening. Check again tomorrow.

**Comment by mgdodge** [comment]
So yeah, it took me a month to get out from under some other work, don't judge... ;)

Anyway, another gist - https://gist.github.com/mgdodge/6e2838d626af3e60d3aaf35fd0faaba5. Basing this again off of my other [vue bug report repo](https://github.com/mgdodge/rollup-plugin-vue-treeshake-bug-vue3), which is now old enough it's accumulated several large vulnerabilities, I ran `npx check-audit` and piped the output to some files for easier comparison. I had the exact same `node_modules` folder in each case, with the intent of seeing what someone would get by upgrading npm and working on an existing project.

I ran the following scenarios:

- `npm-audit-resolver@latest` from npm 6 (baseline)
- `npm-audit-resolver@next` from npm 6
- `npm-audit-resolver@next` from npm 7

vastly different output between npm 6 and 7, understandably, but npm 7 is missing a lot of things that were listed when running under npm 6. This concerns me, as I know several that are not listed are rather widespread. For example, npm 6 shows some warnings that exist in lodash that affected hundreds of packages, and the npm 7 listing is absolutely silent about it.

I've kept all my actual projects back at npm6 for now, as it's become vital for us to be able to automate our audits and catch these types of things as they come up. Sorry about the long turnaround on this.

**Comment by naugtur** [comment]
@mgdodge That's serious.   
Have you checked if npm audit output without the resolver also differs that much? maybe something's wrong with my algorithm for finding paths from top level dependencies and I'm dropping something from the npm7 output?
I'd check it myself, but I'm not sure which branch you used from the repo and I don't want to guess.

**Comment by mgdodge** [comment]
Just ran the tests again, on a new machine, to verify things. Here are the exact steps I followed:

- Clone [repo](https://github.com/mgdodge/rollup-plugin-vue-treeshake-bug-vue3), check out vue 2 branch
- From npm 6, run `npm ci` to install from package-lock.json
- Run `npm i -D npm-audit-resolver@latest`
- Run `npx check-audit` with npm 6, piping output to a file

Then, I switched to npm 7 (leaving same node_modules and package-lock.json files in place) and did the following:
- Run `npm i -D npm-audit-resolver@next`
- (Running this caused the package-lock.json to be rewritten) - as a user migrating to npm 7, I would expect this, and it's a necessary step for someone running a version of npm-audit-resolver that doesn't support npm 7 anyway.
- Run `npx check-audit` with npm 7, piping output to a file

Then, just for good measure, I ran `npm i -g npm@6` to roll back to npm 6, and ran `npx check-audit` one last time on npm 6. You will note that the first time on npm 6, `npm-audit-resolver@latest` was used, and the second time on npm6, `npm-audit-resolver@next` was used.

The results of this second test are the same as I reported above - both npm 6 runs show all the lodash/postcss/browserslist errors, while npm 7 is silent about all of them.

As I said, I did this all on a fresh machine, so you should be able to duplicate things. Also, if system information is important, I am running this on an Ubuntu 20 WSL installation, inside the native linux filesystem. I am using nvm to manage the versions of node installed, and did all this in the latest lts version available.

**Comment by naugtur** [comment]
npm7 audit itself does list more vulnerabilities. Looks like there's a bug in my algorithm that's selecting the dependency paths pointing to vulnerabilities. I might be loosing something on the way. 
Gotta convert your example to a test case now.

**Comment by naugtur** [comment]
Found it. Thanks for the testcase @mgdodge 
Using the example of lodash - lodash is not listed in `npm ls --depth 0` but it doesn't have a `via` field in its vulnerability definition, so it's not reachable. No `via` seems like a bug. so there it is https://github.com/npm/cli/issues/3534

**Comment by mgdodge** [comment]
Okay, ran the new version and can confirm that it is now finding the lodash issues and others. I have noticed now that the `fix` option in `npm-audit-resolver` does nothing except update the `audit-resolve.json` file and tell me to run `npm audit fix` to fix the actual issues - but doing so does nothing. Might be the same thing [reported here](https://github.com/npm/cli/issues/3472). Npm audit is such a mess now!

Appreciate all the work you are doing on this. Npm 7 is still unusable, though, since their own native command does nothing. Any ideas?

**Comment by naugtur** [comment]
I fixed resolver by working around what seems like a bug in npm. Then I used your repo to report a bug in npm. A bit over an hour ago I mentioned it in npm meeting. 

See here https://github.com/npm/cli/issues/created_by/naugtur

**Comment by adon-at-work** [comment]
thanks for all the hardwork. may I ask if there's any update?

**Comment by naugtur** [comment]
Thanks for the polite nudge. (warning: not all maintainers like it :P )
You can install from the `next` tag and pretty much use it, just without the 'fix' option. I got stuck on getting that to work but pushed one change into npm internals that helps with some of the issues

Meanwhile I considered rebuilding the whole thing on top of arborist (the engine behind npm) which would take away many of the issues but also risk losing yarn support.
Late last year npm folks told me what they're planning for 2022 and it'd make my work much easier.

Current state - on hold, life's been taking more time than expected. Hoping to get back to it by the end of winter.

**Comment by joebowbeer** [comment]
It would be great if the next `npm-audit-resolver` were ready in time for those who need to move off of Node12 when it reaches EOL on 4/30/2022

**Comment by naugtur** [comment]
Good point! 
When I get back to it and scaffold things I'll ping y'all for help 😊

**Comment by naugtur** [comment]
I solved outstanding stuff (mostly by deciding to drop the fix option, because it was too complicated to handle it in any form with marking stuff as fixed) and 3.0.0-7 is a decent release candidate.  
Tests welcome!

**Comment by naugtur** [comment]
lol, I'm missing a test :D BRB

**Comment by naugtur** [comment]
@Fryuni works for me. Can you help me reproduce?

**Comment by ronjouch** [comment]
> I solved outstanding stuff (mostly by deciding to drop the fix option, because it was too complicated to handle it in any form with marking stuff as fixed) and 3.0.0-7 is a decent release candidate. Tests welcome!

@naugtur I have been successfully running `-6` on several projects before, I upgraded one to `-7`, and so far so good for me, feels ready for release. Node 14 on Linux, Node 6 + 8, Ubuntu + AL2.

- [x] Exits 1 on missing/unfixed ignores
- [x] Exits 0 when ignoring the right things
- [x] Exits 0 when I add a [`"reason"`](https://github.com/naugtur/npm-audit-resolver/issues/23)
- [x] `resolve-audit` assistant works fine

**Comment by Fryuni** [comment]
Update from our use too. About 15 projects, both front and back all updated to `-7`. Besides the caching bug on GH actions for whatever reason, everything is working perfectly.

We use the `setup-node@v3` action with node 16 to run on CI. Devs are all running some version of node 16 locally too and so far no problem

**Comment by Fryuni** [comment]
We haven't updated any yet, but we added some fixes and temporary ignores and it worked fine.

**Comment by naugtur** [comment]
The latest version doesn't support adding fixes. I'm confused :D

**Comment by ronjouch** [comment]
@naugtur waddayamean by “adding fixes”. I do confirm that I was able in `3.0.0-7` to:

1. `./node_modules/bin/check-audit` and exit non-0
2. `./node_modules/bin/resolve-audit`, add a permanent exception
3. rerun `./node_modules/bin/check-audit` and get an exit 0

**Comment by naugtur** [comment]
I meant I removed the option to apply the npm audit fix on a package. You sounded like you used it 😉

Thanks for clarifying

**Comment by naugtur** [comment]
oh and BTW. @ronjouch - if you run the commands inside n npm script (`npm run test` etc) you can use the command names directly. You can also do `npx resolve-audit` or `yarn resolve-audit` dependeing ehich one you use and it'll work.

**Comment by Fryuni** [comment]
By adding fixes I meant resolving the problems. We run `npm audit fix` directly and then `resolve-audit` to ignore the ones that we inspect to be safe. I forgot there was a `fix` option inside `resolve-audit`, sorry for the confusion @naugtur

**Comment by bytestream** [comment]
@naugtur I'm using `3.0.0-7`. Where do the IDs come from in the `check-audit` output?
```console
$ npx check-audit --production --json
[
    {
        "id": 1045707,
        "name": "minimist",
        "title": "Prototype Pollution in minimist",
        "url": "https://github.com/advisories/GHSA-vh95-rmgr-6w4m",
        "severity": "moderate",
        "range": "<0.2.1",
        "fixAvailable": false,
        "paths": [
            "selectize>sifter>optimist>minimist"
        ],
        "resolutions": [
            {
                "path": "selectize>sifter>optimist>minimist",
                "resolution": "none"
            }
        ]
    }
]
```

The IDs seem to frequently change, for example yesterday it was `1035259` which I added to `audit-resolve.json`:
```json
    "1035259|selectize>sifter>optimist>minimist": {
      "decision": "ignore",
      "madeAt": 1646665320617
    }
```

**Comment by naugtur** [comment]
Hmm. That's weird. 

My code is not producing any information, it must be coming from npm.
Can you share the pkgjson and pkglock with me? Can be privately over email too

**Comment by bytestream** [comment]
I've sent an email to your gmail

**Comment by Fryuni** [comment]
Heads up, this is also happening for some of our devs. It might be a coincidence for our case that, but it is only happening on recently reported vulnerability. I have not checked deeply so this is only speculation:
The `id` field is coming from the `source` field on the audit json response. Could it be related to the change from the npm advisory database to the GH Advisory database, this happened in October, but only recently (I think a few days/weeks ago) did they actually take down the npm advisories page and set it as a redirect to the GH database

**Comment by naugtur** [comment]
@ruyadorno @nlf Can you confirm that observation?

**Comment by naugtur** [comment]
The issue with IDs should now be behind us, I'm getting ready to finish off v3 and get it published. 
Who's excited? :D

**Comment by joebowbeer** [comment]
Is npm7 still maintained? The last release (npm v7.24.2) was 1 year ago.

Given that Node 16 enters maintenance phase in 60 days, and even it ships with npm v8, is there any need to support npm v7?

**Comment by naugtur** [comment]
Edited the title. @joebowbeer

**Comment by mgdodge** [comment]
@naugtur , with npm 9 just being released, I'm curious if they've messed with the audit format again. I think historically there's enough data here to verify things, curious if you've got an eye on that or not.

**Comment by naugtur** [comment]
My eyes were elsewhere mostly. Thanks for bringing it up.
I was hoping to release v3 for Christmas so gotta check what they did 😁

**Comment by naugtur** [comment]
@mgdodge seems to be working with npm9

**Comment by joebowbeer** [comment]
I notice that Node 18.14.0 comes bundled with npm v9

```
nvm use 18.13
Now using node v18.13.0 (npm v8.19.3)
```

```
nvm use 18.14
Now using node v18.14.0 (npm v9.3.1)
```

The npm v9 lockfileVersion is still 2, right?

https://docs.npmjs.com/cli/v9/configuring-npm/package-lock-json#lockfileversion

**Comment by naugtur** [comment]
audit resolver doesn't depend on the lockfile format, that's the whole reason behind going through the hassle of calling the cli commands - they're the most stable API surface available.

**Comment by joebowbeer** [comment]
Can this be closed as completed?

**Comment by naugtur** [comment]
I think it finally can, with some ugly fixes for npm6 tests...
